internal/consensus/reactor.go

```diff
@@ -1021,19 +1022,35 @@ func (r *Reactor) processPeerUpdate(ctx context.Context, peerUpdate p2p.PeerUpda
 			// when the peer is removed. We also set the running state to ensure we
 			// do not spawn multiple instances of the same goroutines and finally we
 			// set the waitgroup counter so we know when all goroutines have exited.
-			ps.broadcastWG.Add(3)
 			ps.SetRunning(true)
 
-			// start goroutines for this peer
-			go r.gossipDataRoutine(ctx, ps)
-			go r.gossipVotesRoutine(ctx, ps)
-			go r.queryMaj23Routine(ctx, ps)
-
-			// Send our state to the peer. If we're block-syncing, broadcast a
-			// RoundStepMessage later upon SwitchToConsensus().
-			if !r.waitSync {
-				go func() { _ = r.sendNewRoundStepMessage(ctx, ps.peerID) }()
+			startPeerRoutine := func() {
+				// do nothing if we or the peer has
+				// stopped while we've been waiting.
+				if !ps.IsRunning() || ctx.Err() != nil {
+					return
+				}
+				ps.broadcastWG.Add(3)
+				// start goroutines for this peer
+				go r.gossipDataRoutine(ctx, ps)
+				go r.gossipVotesRoutine(ctx, ps)
+				go r.queryMaj23Routine(ctx, ps)
+
+				// Send our state to the peer. If we're block-syncing, broadcast a
+				// RoundStepMessage later upon SwitchToConsensus().
+				if !r.WaitSync() {
+					go func() { _ = r.sendNewRoundStepMessage(ctx, ps.peerID) }()
+				}
 			}
+			select {
+			case <-ctx.Done():
+				return
+			case <-r.readySignal:
+				startPeerRoutine()
+			default:
+				go startPeerRoutine()
```

Comment:
Hmm. Doesn't this obviate the ready signal, since we're not waiting for it?

Reply:
the process peer update operation doesn't want to block, so we have to do something to avoid blocking. I can rejigger things

Reply:
It seems like maybe what we want here is, say

```go
go func() {
  <-r.readySignal
  startpeerRoutine()
}()
```

Reply:
yup yup, that's more or less what I did

Reply:
eyeballing it thought, it looks like we're calling Wait() and Add from different threads in a way that could easily explode. I think given contexts its safe to just delete the wait group, so I'll do that here too.